src/daemon/daemon.cpp

```diff
@@ -281,25 +283,35 @@ auto fetch_image_for(const std::string& name, const mp::FetchType& fetch_type, m
 
 auto validate_create_arguments(const mp::LaunchRequest* request)
 {
-    mp::LaunchError launch_error;
+    auto mem_size = request->mem_size();
+    auto disk_space = request->disk_space();
+    auto instance_name = request->instance_name();
+    auto option_errors = mp::LaunchError{};
 
-    if (!request->disk_space().empty() && !mp::utils::valid_memory_value(QString::fromStdString(request->disk_space())))
-    {
-        launch_error.add_error_codes(mp::LaunchError::INVALID_DISK_SIZE);
-    }
+    const auto opt_mem_size = mp::utils::in_bytes(mem_size.empty() ? "1G" : mem_size);
+    const auto opt_disk_space = mp::utils::in_bytes(disk_space.empty() ? "5G" : disk_space);
 
-    if (!request->mem_size().empty() && !mp::utils::valid_memory_value(QString::fromStdString(request->mem_size())))
-    {
-        launch_error.add_error_codes(mp::LaunchError::INVALID_MEM_SIZE);
-    }
+    if (opt_mem_size && *opt_mem_size >= normalized_min_mem)
+        mem_size = mp::utils::in_bytes_string(*opt_mem_size);
+    else
+        option_errors.add_error_codes(mp::LaunchError::INVALID_MEM_SIZE);
 
-    if (!request->instance_name().empty() &&
-        !mp::utils::valid_hostname(QString::fromStdString(request->instance_name())))
-    {
-        launch_error.add_error_codes(mp::LaunchError::INVALID_HOSTNAME);
-    }
+    if (opt_disk_space && *opt_disk_space >= normalized_min_disk)
+        disk_space = mp::utils::in_bytes_string(*opt_disk_space);
+    else
+        option_errors.add_error_codes(mp::LaunchError::INVALID_DISK_SIZE);
+
+    if (!request->instance_name().empty() && !mp::utils::valid_hostname(request->instance_name()))
+        option_errors.add_error_codes(mp::LaunchError::INVALID_HOSTNAME);
 
-    return launch_error;
+    struct CheckedArguments
+    {
+        std::string mem_size;
+        std::string disk_space;
+        std::string instance_name;
+        mp::LaunchError option_errors;
+    } ret{mem_size, disk_space, instance_name, option_errors};
```

Comment:
This doesn't address my concern. There is still no connection between this ordering of params and the order in the consumer. E.g. this will compile, but is wrong:
```
auto [disk_space, mem_size, instance_name, option_errors] = validate_create_arguments(request);
```
This is the root of my dislike of this syntactic sugar.

If the params were all different types, the type system would catch errors. So I'm ok with this syntax being used with care. But here, I think using a struct like this would give more safety. Here's my thinking:
https://pastebin.ubuntu.com/p/RsQ3sf6dZy/

My motivation is that this moves the responsibility for parameter order checking to solely the location the param list is created (i.e. in validate_create_arguments). Not where it is consumed. 

With this struct approach, if someone swaps mem_size and disk_space, they only need to care about reordering those params in the creation of the struct. No other code needs to change, and no bugs created.

Sure, we could go all the way and insist on this at creation:
```
CheckedArguments args;
args.mem_size = mem_size;
args.disk_space = ....
return args;
```
but I think that's unnecessary. I trust that if an author wants to change the CheckedArguments list order, they'll be sure to do the right thing. But I don't trust they'll update all the consumers correctly too.

Is it pretty? No. But IMO it's worth it for safety.

Reply:
Just my $0.02, but that pastebin is what I envisioned too when I made my comments.

Reply:
OK, I will do it that way then, but...

> My motivation is that this moves the responsibility for parameter order checking to solely the location the param list is created (i.e. in validate_create_arguments). Not where it is consumed.

Well, the ordering of public attributes is part of a struct's interface, so it is not being contained. IOW, anyone changing the order would still have to check all clients (think also construction, assignment...). But I understand your concern is that they may forget so it is perhaps safer not to rely on that here.